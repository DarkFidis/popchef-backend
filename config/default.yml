server:
  name: popchef-test
  listen:
    port: 4000
  poweredBy: darkfidis
log:
  level: debug
