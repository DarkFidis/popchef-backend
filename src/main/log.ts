import { Logger } from 'winston'

import { getLogger } from './utils/logger'

const log: Logger = getLogger('popchef-test')

export { log }
