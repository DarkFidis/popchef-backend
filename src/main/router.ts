import {
  json as jsonBodyParserFactory,
  raw as rawBodyParserFactory,
  urlencoded as urlencodedBodyParserFactory,
} from 'body-parser'
import cookieParserFactory from 'cookie-parser'
import corsFactory from 'cors'
import { ApolloServer } from "apollo-server-express";

import { helloWorldMw } from './middlewares/hello-world'
import { RegisterApp } from './types/web-server'
import { toExpressMw } from './utils/helper'
import {userRouter} from "./routes/user.router";
import {gqlErrorFormatter} from "./graphql/formatOkapiErrors";
import {createSchema} from "./graphql/createSchema";

const jsonBodyParserMw = jsonBodyParserFactory()
const urlencodedBodyParserMw = urlencodedBodyParserFactory({ extended: false })
const rawBodyParserMw = rawBodyParserFactory({ limit: '10mb', type: '*/*' })
const cookieParserMw = cookieParserFactory()
const corsMw = corsFactory({ origin: ['http://localhost:4000']})

const registerGqlServer: RegisterApp = async (app) => {
  const schema = await createSchema()
  const apolloServer = new ApolloServer({
    formatError: gqlErrorFormatter,
    schema,
    context: ({ req, res }: any) => ({ req, res })
  })
  await apolloServer.start();
  apolloServer.applyMiddleware({ app })
}

export const registerApp: RegisterApp = async (app) => {
  app.use(cookieParserMw, jsonBodyParserMw, urlencodedBodyParserMw, rawBodyParserMw, corsMw)
  app.use('/api/users', userRouter)
  app.get('/hello', toExpressMw(helloWorldMw))
  await registerGqlServer(app)
}
