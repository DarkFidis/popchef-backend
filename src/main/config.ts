import { resolve } from 'path'

process.env.NODE_CONFIG_DIR = resolve(__dirname, '..', '..', 'config')

import nodeConfig from 'config'

import { version } from '../../package.json'
import { Config } from './types/config'

const env = process.env.NODE_CONFIG_ENV || process.env.NODE_ENV || 'development'
const log = nodeConfig.has('log') ? nodeConfig.get<Config['log']>('log') : { level: 'info' }
const server = nodeConfig.has('server')
  ? nodeConfig.get<Config['server']>('server')
  : { listen: {} }

export const config: Config = {
  env,
  log,
  name: 'popchef-test',
  server,
  version,
}
