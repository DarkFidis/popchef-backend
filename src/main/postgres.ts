import {DataSourceOptions} from "typeorm";
import {PgClient} from "./services/pg-client";
import {log} from "./log";

const dbConfig: DataSourceOptions = {
  type: "postgres",
  host: "localhost",
  port: 5432,
  username: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_NAME,
  entities: ['main/db/entities/*.js'],
  logging: true,
  synchronize: true,
}

export const pgClient = new PgClient(log, dbConfig)
