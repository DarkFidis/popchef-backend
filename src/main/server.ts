import { log } from './log'
import { registerApp } from './router'
import { WebServer } from './services/web-server'
import { WebServerable } from './types/web-server'

const webServer: WebServerable = new WebServer(log, registerApp)

export { webServer }
