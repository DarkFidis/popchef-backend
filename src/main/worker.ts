import { log } from './log'
import { webServer } from './server'
import { Workerable } from './types/worker'

const worker: Workerable = {
  handleSignal: async (name) => {
    log.info(`received ${name} signal : stopping`)
    await worker.shutdown()
  },
  run: async () => {
    const signals: NodeJS.Signals[] = ['SIGINT', 'SIGTERM']
    signals.forEach((signal) => {
      process.on(signal, () => {
        void worker.handleSignal(signal)
      })
    })
    process.on('message', (msg) => {
      log.info(`received message : ${msg}`)
      if (msg === 'shutdown') {
        log.info(`received shutdown message : stopping`)
        void worker.shutdown(0)
      }
    })
    await webServer.init()
    await webServer.start()
    log.info(`/!\\ to stop worker : kill -s SIGTERM ${process.pid}`)
  },
  shutdown: async (exitCode = 1) => {
    try {
      await webServer.stop()
      process.exit(exitCode)
    } catch (err) {
      log.error(err)
      process.exit(1)
    }
  },
}

export = worker
