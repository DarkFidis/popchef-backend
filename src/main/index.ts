import "reflect-metadata"

import { run } from "./worker"

run().catch(err => console.error(err))
