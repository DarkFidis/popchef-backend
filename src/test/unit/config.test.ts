import { IConfig } from 'config'
import { resolve } from 'path'

describe('config unit tests', () => {
  const envVarNames = [
    'BASE_URL',
    'LOG_LEVEL',
    'NODE_CONFIG_ENV',
    'NODE_ENV',
    'PORT',
    'REDIS_HOST',
    'REDIS_PORT',
  ]
  const envMapOrig: {
    [key: string]: string | undefined
  } = {}
  beforeAll(() => {
    envVarNames.forEach((name) => {
      envMapOrig[name] = process.env[name]
    })
  })
  beforeEach(() => {
    envVarNames.forEach((name) => {
      delete process.env[name]
    })
    jest.doMock('../../../package.json', () =>
      Object.assign({}, jest.requireActual('../../../package.json'), {
        version: '0.0.1-fakeversion',
      }),
    )
  })
  afterEach(() => {
    jest.resetModules()
  })
  afterAll(() => {
    envVarNames.forEach((name) => {
      envMapOrig[name] = process.env[name]
    })
  })
  describe('custom env vars', () => {
    test.each([
      [undefined, undefined, undefined, undefined, undefined],
      ['http://mybaseurl', undefined, undefined, undefined, undefined],
      ['http://mybaseurl', 3333, undefined, undefined, undefined],
      ['http://mybaseurl', 3333, 'debug', undefined, undefined],
      ['http://mybaseurl', 3333, 'debug', 'redisHost', undefined],
      ['http://mybaseurl', 3333, 'debug', 'redisHost', 6379],
    ])(
      'should return custom config given BASE_URL=%p, PORT=%p, LOG_LEVEL=%p, REDIS_HOST=%p, REDIS_PORT=%p',
      (
        baseUrl?: string,
        port?: number,
        logLevel?: string,
        redisHost?: string,
        redisPort?: number,
      ) => {
        // Given
        if (baseUrl) {
          process.env.BASE_URL = baseUrl
        }
        if (port) {
          process.env.PORT = `${port}`
        }
        if (logLevel) {
          process.env.LOG_LEVEL = logLevel
        }
        if (redisHost) {
          process.env.REDIS_HOST = redisHost
        }
        if (redisPort) {
          process.env.REDIS_PORT = `${redisPort}`
        }
        // When
        const config = require('../../main/config')
        // Then
        expect(config).toMatchSnapshot()
      },
    )
  })
  describe('hacked config', () => {
    let nodeConfig: jest.Mocked<IConfig>
    beforeAll(() => {
      jest.resetModules()
      jest.doMock('config')
      nodeConfig = require('config')
    })
    test('should return config with defaults given no config is provided for log, server', () => {
      // Given
      nodeConfig.has.mockReturnValue(false)
      // When
      const { config } = require('../../main/config')
      // Then
      expect(config).toHaveProperty('server')
      expect(config).toMatchSnapshot()
    })
  })
  describe('NODE_CONFIG_DIR', () => {
    test('should set NODE_CONFIG_DIR env var to /config from project base dir', () => {
      // Given
      const baseDir = resolve(__dirname, '..', '..', '..')
      // When
      require('../../main/config')
      // Then
      expect(process.env).toHaveProperty('NODE_CONFIG_DIR', resolve(baseDir, 'config'))
    })
  })
})
