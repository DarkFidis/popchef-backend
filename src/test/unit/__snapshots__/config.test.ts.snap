// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config unit tests custom env vars should return custom config given BASE_URL="http://mybaseurl", PORT=3333, LOG_LEVEL="debug", REDIS_HOST="redisHost", REDIS_PORT=6379 1`] = `
Object {
  "config": Object {
    "env": "development",
    "log": Object {
      "level": "debug",
    },
    "name": "popchef-test",
    "server": Object {
      "listen": Object {
        "port": 4000,
      },
      "name": "popchef-test",
      "poweredBy": "darkfidis",
    },
    "version": "0.0.1-fakeversion",
  },
}
`;

exports[`config unit tests custom env vars should return custom config given BASE_URL="http://mybaseurl", PORT=3333, LOG_LEVEL="debug", REDIS_HOST="redisHost", REDIS_PORT=undefined 1`] = `
Object {
  "config": Object {
    "env": "development",
    "log": Object {
      "level": "debug",
    },
    "name": "popchef-test",
    "server": Object {
      "listen": Object {
        "port": 4000,
      },
      "name": "popchef-test",
      "poweredBy": "darkfidis",
    },
    "version": "0.0.1-fakeversion",
  },
}
`;

exports[`config unit tests custom env vars should return custom config given BASE_URL="http://mybaseurl", PORT=3333, LOG_LEVEL="debug", REDIS_HOST=undefined, REDIS_PORT=undefined 1`] = `
Object {
  "config": Object {
    "env": "development",
    "log": Object {
      "level": "debug",
    },
    "name": "popchef-test",
    "server": Object {
      "listen": Object {
        "port": 4000,
      },
      "name": "popchef-test",
      "poweredBy": "darkfidis",
    },
    "version": "0.0.1-fakeversion",
  },
}
`;

exports[`config unit tests custom env vars should return custom config given BASE_URL="http://mybaseurl", PORT=3333, LOG_LEVEL=undefined, REDIS_HOST=undefined, REDIS_PORT=undefined 1`] = `
Object {
  "config": Object {
    "env": "development",
    "log": Object {
      "level": "debug",
    },
    "name": "popchef-test",
    "server": Object {
      "listen": Object {
        "port": 4000,
      },
      "name": "popchef-test",
      "poweredBy": "darkfidis",
    },
    "version": "0.0.1-fakeversion",
  },
}
`;

exports[`config unit tests custom env vars should return custom config given BASE_URL="http://mybaseurl", PORT=undefined, LOG_LEVEL=undefined, REDIS_HOST=undefined, REDIS_PORT=undefined 1`] = `
Object {
  "config": Object {
    "env": "development",
    "log": Object {
      "level": "debug",
    },
    "name": "popchef-test",
    "server": Object {
      "listen": Object {
        "port": 4000,
      },
      "name": "popchef-test",
      "poweredBy": "darkfidis",
    },
    "version": "0.0.1-fakeversion",
  },
}
`;

exports[`config unit tests custom env vars should return custom config given BASE_URL=undefined, PORT=undefined, LOG_LEVEL=undefined, REDIS_HOST=undefined, REDIS_PORT=undefined 1`] = `
Object {
  "config": Object {
    "env": "development",
    "log": Object {
      "level": "debug",
    },
    "name": "popchef-test",
    "server": Object {
      "listen": Object {
        "port": 4000,
      },
      "name": "popchef-test",
      "poweredBy": "darkfidis",
    },
    "version": "0.0.1-fakeversion",
  },
}
`;

exports[`config unit tests hacked config should return config with defaults given no config is provided for log, server 1`] = `
Object {
  "env": "development",
  "log": Object {
    "level": "info",
  },
  "name": "popchef-test",
  "server": Object {
    "listen": Object {},
  },
  "version": "0.0.1-fakeversion",
}
`;
